### DISCUSSION TITLE
Lazy-populate data

### DISCUSSION THREAD

**Discussion by danilopolani** [squirephp/squire#4] [Ideas]
It's not really an issue, but a suggestion: it would be nice to not have every single file already populated in the repository but, instead, let the user download their desired data. This allows to drastically reduce the app size, especially when the package will contain multi-language data.

For example the `airports.csv` is **3.26MB** and it's a lot. I thought a solution like this would help.

### Solution 1

1. User can download the data:
```shell
$ php artisan squire:download {resources*}
$ php artisan squire:download airports
```

2. The `squire:download` command will check if the file exists and if it's signature is not the same as the latest version available. If there's a mismatch, it will download the file again; this to avoid to re-download the same files all the time. 
2. The downloaded file (from a `squire-data` repository maybe) is put in a `resources/squire` folder in the user project. 
3. The Model loads the data from `resource_path('squire/airports.csv')`

And it would fit with different languages too:

```shell
$ php artisan squire:download {resources*} {--locale}
```

### Solution 2

Everything like solution 1, but instead of passing the desired resources and languages in the command, the package could create a `config.squire.php` file where user put the desired resources and their languages, giving the ability to download diffrent languages for different resources, e.g. `Airports -> en, Countries -> en, it`.

Then the command would be just `php artisan squire:download`.

---

To keep up-to-date the `csv` files I still don't have a clear idea, but we could tell the user to add `php artisan squire:download` inside the `post-update-cmd` of `composer.json` or just to manually pull the data sometimes.

**Comment by danharrin** [comment]
How about having a modular system based on multiple Composer packages? Each model is contained within its own package.

**Comment by danilopolani** [comment]
It could work, tbh my biggest fear is mainly about multi languages, because if a package `airports` provides 15 languages and it will automatically download all of them, you will have a dependency 48MB heavy

**Comment by danharrin** [comment]
From v2, Squire will be split into multiple composer packages. Each will contain a translation for just one model.

For example, to use the `Squire\Models\Country` model in English and French:

```
composer require squirephp/country-en squirephp/country-fr
```

All translations are easily updated, the same as you would with any other package.

**Comment by danilopolani** [comment]
Huge, thanks and great work!

**Comment by danharrin** [comment]
All models (and multiple translations) in v2.0.0 are provided by different packages.
